### PR TITLE
get datums using pallas observer

### LINKED ISSUE #1315: Enhance Mithril/Cardano node communication
## Why
Today, the **Mithril signer and aggregator** communicate with the **Cardano node** with the **Cardano cli** which acts as a proxy for providing some of the information essential to running a **Mithril network**. This communication is not optimal, and we could communicate directly with the **Cardano node** instead.

## What
The **Mithril signer and aggregator** (will) rely on some information that are provided by the **Cardano node** :
- Epoch
- Epoch Stake Distribution
- KES Period
- UTxO of an address (TxDatum in particular)
- Cardano node version (will)
- Protocol Parameters (will)
- ...

The signer and aggregator rely on an a common [`ChainObserver`](https://github.com/input-output-hk/mithril/blob/main/mithril-common/src/chain_observer/interface.rs#L27) trait:
```rust
pub trait ChainObserver: Sync + Send {
    /// Retrieve the datums associated to and address
    async fn get_current_datums(
        &self,
        address: &ChainAddress,
    ) -> Result<Vec<TxDatum>, ChainObserverError>;

    /// Retrieve the current epoch of the Cardano network
    async fn get_current_epoch(&self) -> Result<Option<Epoch>, ChainObserverError>;

    /// Retrieve the current stake distribution of the Cardano network
    async fn get_current_stake_distribution(
        &self,
    ) -> Result<Option<StakeDistribution>, ChainObserverError>;

    /// Retrieve the KES period of an operational certificate
    async fn get_current_kes_period(
        &self,
        _opcert: &OpCert,
    ) -> Result<Option<KESPeriod>, ChainObserverError> {
        Ok(None)
    }
}
```

That is currently implemented by the [`CliObserver`](https://github.com/input-output-hk/mithril/blob/main/mithril-common/src/chain_observer/cli_observer.rs#L387).

We would like to have a new implementation `CardanoObserver` of this trait that:
- removes the usage of the Cardano cli 
- uses a direct communication with the Cardano node through the **socket** it exposes and using **mini-protocols**

A good candidate library to implement this is [`pallas`](https://crates.io/crates/pallas) and [`pallas-network`](https://crates.io/crates/pallas-network) that already has an implementation of the mini-protocols.

**Note**:
The signer and aggregator nodes already have a defined configuration parameter with the path of the Cardano node socket `cardano_node_socket_path`:
- https://mithril.network/doc/manual/developer-docs/nodes/mithril-aggregator#configuration-parameters
- https://mithril.network/doc/manual/developer-docs/nodes/mithril-signer#configuration-parameters

## How

> Draft of how this could be implemented 

- [ ] Evaluate if `pallas` can give access to the aforementioned information from the Cardano node (all of subset of them)
- [ ] Implement the `CardanoObserver` for the current version of the `ChainObserver` trait without relying on the Cardano cli
- [ ] Implement the usage of this new `CardanoObserver` and activate the dependency with a specific configuration (e.g. by adding a new adapter configuration parameter, a bit like the `era_reader_adapter_type` in [signer](https://mithril.network/doc/manual/developer-docs/nodes/mithril-signer#configuration-parameters) and [aggregator](https://mithril.network/doc/manual/developer-docs/nodes/mithril-aggregator#configuration-parameters))
- [ ] If more information than those available in the current version of the `ChainObserver` trait, add them (with blanket implementation to not break current `CliObserver`)? 
- [ ] Document the modifications in the [Mithril User manual](https://mithril.network/doc/manual/welcome)
- [ ] TBD

- Unit tests (mocks) should be implemented
- [End to end tests](https://github.com/input-output-hk/mithril/tree/main/mithril-test-lab/mithril-end-to-end#build-and-run-the-test-suite) should be working consistently
- All checks and tests should be green in the [CI](https://github.com/input-output-hk/mithril/actions/workflows/ci.yml) (documentation, linting, ...)

### PR DESCRIPTION
**Pull request by falcucci**
## Content

Includes:
- get current datums using the new pallas observer;
- brings all compatibilities with the previous cardano-cli format;
- uses the current cardano-cli as a default fallback for the remaining methods;

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested

## Comments

The automatic fallback mode has been implemented to make sure we progressively rollout the new integration.
 

## Issue(s)

fixes #1315

### CODE REVIEW

File: mithril-common/Cargo.toml [hunk 1]
```diff
@@ -74,7 +77,7 @@ mithril-stm = { path = "../mithril-stm", version = "0.3", default-features = fal
 [dev-dependencies]
 criterion = { version = "0.5.1", features = ["html_reports", "async_tokio"] }
 mockall = "0.11.4"
-pallas-codec = "0.20.0"
+pallas-crypto = { git = "https://github.com/falcucci/pallas.git", branch = "any-cbor-inner-pub" }
```
Comment (falcucci) [inline mithril-common/Cargo.toml:80]: must be ignored. we are going to bump it correctly soon

File: mithril-common/src/chain_observer/model.rs [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+#[derive(Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum Metadatum {
+    Datum(Constr<Metadatum>),
+    Int(Int),
+    Bytes(String),
+    Text(String),
+    List(Vec<Metadatum>),
+    Map(KeyValuePairs<Metadatum, Metadatum>),
+}
```
Comment (scarmuega) [inline mithril-common/src/chain_observer/model.rs:27]: there's no need to replicate this, `pallas::primitives` already has all of these structures defined and with the corresponding CBOR codec.

Comment (falcucci) [reply mithril-common/src/chain_observer/model.rs]: done.

**Comment by falcucci** [comment]
@jpraynaud from now on we can review from mithril side. 

we also need to bump the new pallas version asap but regardning to that just ignore the `check` step from now.

**Comment by falcucci** [comment]
the version has been bumped now.

File: mithril-common/Cargo.toml [hunk 3]
```diff
@@ -33,8 +33,12 @@ kes-summed-ed25519 = { version = "0.2.1", features = [
     "serde_enabled",
     "sk_clone_enabled",
 ] }
+minicbor = { version = "0.19", features = ["std", "half", "derive"] }
```
Comment (jpraynaud) [inline mithril-common/Cargo.toml:36]: It looks like a newer version is available:
```suggestion
minicbor = { version = "0.20", features = ["std", "half", "derive"] }
```

File: mithril-common/Cargo.toml [hunk 4]
```diff
@@ -33,8 +33,12 @@ kes-summed-ed25519 = { version = "0.2.1", features = [
...
 nom = "7.1.3"
-pallas-network = { version = "0.20.0", optional = true }
+pallas-addresses = "0.21.0"
+pallas-codec = "0.21.0"
+pallas-network = { version = "0.21.0", optional = true }
+pallas-primitives = "0.21.0"
```
Comment (jpraynaud) [inline mithril-common/Cargo.toml:41]: We need to have `pallas` crates behind the `fs` feature:
```suggestion
pallas-addresses = { version = "0.21.0", optional = true }
pallas-codec = { version = "0.21.0", optional = true }
pallas-network = { version = "0.21.0", optional = true }
pallas-primitives = { version = "0.21.0", optional = true }
```
And the feature would be defined as (line 101):
```
fs = [
    "tokio/fs",
    "tokio/process",
    "pallas-network",
    "pallas-addresses",
    "pallas-primitives",
    "pallas-codec",
]
```

File: mithril-common/src/chain_observer/model.rs [hunk 5]
```diff
@@ -1,12 +1,42 @@
 use anyhow::anyhow;
-use serde::Serialize;
+use pallas_codec::minicbor::{decode, Decode, Decoder};
+use pallas_primitives::{conway::PlutusData, ToCanonicalJson};
```
Comment (jpraynaud) [inline mithril-common/src/chain_observer/model.rs:3]: Do we want to target specifically `conway` era?

File: mithril-common/src/chain_observer/pallas_observer.rs
Comment (jpraynaud) [inline mithril-common/src/chain_observer/pallas_observer.rs:1]: Could you remove all the `unwrap` and `expect` in all the non test code, and replace them with appropriate errors?

File: mithril-common/src/chain_observer/model.rs [hunk 6]
```diff
@@ -1,12 +1,42 @@
...
+}
+
+impl ToCanonicalJson for Datum {
+    fn to_json(&self) -> serde_json::Value {
+        self.0.to_json()
+    }
+}
+
+/// Inspects the given bytes and returns a decoded `R` instance.
+pub fn inspect<R>(inner: Vec<u8>) -> R
```
Comment (jpraynaud) [inline mithril-common/src/chain_observer/model.rs:30]: Why didn't you implement the `TryFrom` trait here?

Alos, this should not panic, and return an error instead:
```suggestion
pub fn try_inspect<R>(inner: Vec<u8>) -> StdResult<R>
```

File: mithril-common/src/chain_observer/pallas_observer.rs [hunk 7]
```diff
@@ -78,6 +83,81 @@ impl PallasChainObserver {
...
+    }
+
+    /// Maps the given `UTxOByAddress` instance to Datums.
+    fn map_datums(&self, transaction: UTxOByAddress) -> Datums {
+        transaction
+            .utxo
+            .iter()
+            .filter(|(_, utxo)| utxo.inline_datum.is_some())
+            .map(|(_, utxo)| self.serialize_datum(utxo))
+            .collect()
```
Comment (jpraynaud) [inline mithril-common/src/chain_observer/pallas_observer.rs:115]: You could have probably used `FilterMap` here:
```suggestion
        transaction
            .utxo
            .iter()
            .filter_map(|(_, utxo)| utxo.inline_datum.map(|d| self.serialize_datum(d)))
            .collect()
```

File: mithril-common/src/chain_observer/pallas_observer.rs [hunk 8]
```diff
@@ -78,6 +83,81 @@ impl PallasChainObserver {
...
+        statequery
+            .acquire(None)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver Failed to acquire statequery")?;
+
+        let era = queries_v16::get_current_era(statequery)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver Failed to get current era")?;
```
Comment (jpraynaud) [inline mithril-common/src/chain_observer/pallas_observer.rs:145]: Can you fix this typo (which is also in `get_epoch` function)?
```suggestion
            .with_context(|| "PallasChainObserver failed to get current era")?;
```

File: mithril-common/src/chain_observer/pallas_observer.rs [hunk 9]
```diff
@@ -158,13 +245,46 @@ impl ChainObserver for PallasChainObserver {
...
-    use pallas_network::miniprotocols::localstate::{self, ClientQueryRequest};
+    use pallas_codec::utils::{AnyCbor, AnyUInt, KeyValuePairs, TagWrap};
+    use pallas_crypto::hash::Hash;
+    use pallas_network::miniprotocols::localstate::{self, queries_v16::Value, ClientQueryRequest};
     use tokio::net::UnixListener;
 
     use super::*;
     use crate::{chain_observer::test_cli_runner::TestCliRunner, CardanoNetwork};
 
+    fn get_utxo_mock() -> UTxOByAddress {
```
Comment (jpraynaud) [inline mithril-common/src/chain_observer/pallas_observer.rs:256]: This could be named differently as it is not really a mock:
```suggestion
    fn get_fake_utxo_by_address() -> UTxOByAddress {
```

**Review by jpraynaud — CHANGES_REQUESTED** [review]
Thanks for the good work @falcucci :+1: 
I have left few comments and questions below.

Also, could you avoid using any `panic`, `unwrap` and `expect` in all the non test code, and replace them with appropriate errors?

File: mithril-common/src/chain_observer/model.rs (hunk 5 above)
Comment (falcucci) [reply mithril-common/src/chain_observer/model.rs]: this is actually the plutus structure which doesnt really depends on the era, it has its own version and we follow the latest available.

The plutus contract for inline datum follows the same for any era by now.

The import is needed from premitives to avoid repeating ourselves.

The root of this import should be alonzo tho, where it is really implemented by pallas-primitives. Just have updated it.

Comment (falcucci) [reply mithril-common/src/chain_observer/model.rs]: reference: https://github.com/IntersectMBO/plutus/blob/master/plutus-core/plutus-core/src/PlutusCore/Data.hs

File: mithril-common/Cargo.toml (hunk 4 above)
Comment (falcucci) [reply mithril-common/Cargo.toml]: just to make it documented,

we got errors to compile to warm after these changes.

@jpraynaud suggested to create a conditional macro to fix the build errors, resulting as:

```

macro_rules! cfg_fs {
    ($($item:item)*) => {
        $(
            #[cfg(feature = "fs")]
            #[cfg_attr(docsrs, doc(cfg(feature = "fs")))]
            $item
        )*
    }
}
```
thanks

**Comment by falcucci** [comment]
@jpraynaud thank you! all addresses upon request has been applied.

File: mithril-common/src/chain_observer/model.rs [hunk 10]
```diff
@@ -7,6 +8,43 @@ use thiserror::Error;
...
+    #[serde(rename_all = "lowercase")]
+    pub struct Datum(pub PlutusData);
+
+    impl ToCanonicalJson for Datum {
+        fn to_json(&self) -> serde_json::Value {
+            self.0.to_json()
+        }
+    }
+
+    impl<'a, C> minicbor::Decode<'a, C> for Datum {
```
Comment (jpraynaud) [inline mithril-common/src/chain_observer/model.rs:25]: Why not use imports for `minicbor`?
IMO it's a bit cumbersome to repeat it in these functions.

File: mithril-common/src/chain_observer/pallas_observer.rs [hunk 11]
```diff
@@ -1,7 +1,11 @@
 use anyhow::{anyhow, Context};
 use async_trait::async_trait;
+use pallas_addresses::Address;
+use pallas_codec::utils::{Bytes, CborWrap, TagWrap};
 use pallas_network::facades::NodeClient;
+use pallas_network::miniprotocols::localstate::queries_v16::{Addr, Addrs, UTxOByAddress, Values};
 use pallas_network::miniprotocols::localstate::{queries_v16, Client};
```
Comment (jpraynaud) [inline mithril-common/src/chain_observer/pallas_observer.rs:7]: This could probably be rewritten as:
```suggestion
use pallas_network::{
    facades::NodeClient,
    miniprotocols::localstate::queries_v16::{self, Addr, Addrs, Client, UTxOByAddress, Values}
};
```

File: mithril-common/src/chain_observer/pallas_observer.rs
Comment (jpraynaud) [inline mithril-common/src/chain_observer/pallas_observer.rs:1]: There are still some occurrences of `PallasChainObserver Failed` in the code form previous PR.
Could you also fix the case in it?

**Review by jpraynaud — APPROVED** [review]
LGTM :+1: 

Thanks for the good work @falcucci! I just left few more comments below.

Also, we probably need to wait until the end to end test is upgraded with the capability to read era data from the chain before we can merge the PR.

**Comment by falcucci** [comment]
@jpraynaud thank you, 

looks like the #1423 has been merged, so i'll pull it and update the crates to make it ready for a merge.

**Review by ghubertpalo — APPROVED** [review]

**Review by dlachaumepalo — APPROVED** [review]
LGTM, thanks for the work @falcucci 🚀

File: mithril-common/Cargo.toml [hunk 12]
```diff
@@ -33,8 +33,12 @@ kes-summed-ed25519 = { version = "0.2.1", features = [
     "serde_enabled",
     "sk_clone_enabled",
 ] }
+minicbor = { version = "0.20", features = ["std", "half", "derive"] }
```
Comment (Alenar) [inline mithril-common/Cargo.toml:36]: Can this dependency be optional since it's only used behind the `fs` flag ?

Comment (falcucci) [reply mithril-common/Cargo.toml]: for sure, done.

**Review by Alenar — APPROVED** [review]

**Comment by jpraynaud** [comment]
Hi @falcucci! 

Can you rebase your branch on the `main` branch as the new `devnet` that tests the UTxO retrieval is now available?

**Comment by falcucci** [comment]
@jpraynaud hey, done!

**Comment by falcucci** [comment]
@jpraynaud I've just fixed a possible serialization compatibility with the `thales` era (I suppose), the problem was that we are receiving the `inline_datum` wrapped by a plus Tag(121) which wasn't matching the expected struct.

From now on I am not sure if we have an issue in the CI side. It needs further investigation but it worked on my machine as you can see.

TLDR; the error we are getting now I could reproduce and it was related to fetch the epoch settings at 

```
http://localhost:8080/aggregator/epoch-settings
```

I just restarted the tests and it worked again as expected but restarting the CI step doesnt have the same results.

CleanShot 2024-01-23 at 02 07 55@2x

**Comment by jpraynaud** [comment]
> @jpraynaud I've just fixed a deserialization issue to consume the inline datum.
> 
> From now on I am not sure if we have an issue in the CI side. It needs further investigation but it worked on my machine as you can see in the pic bellow.
> 
> TLDR; the error we are getting now I could reproduce and it was related to fetch the epoch settings at
> 
> ```
> http://localhost:8080/aggregator/epoch-settings
> ```
> 
> At my machine I just restarted the tests and it worked again as expected but restarting the CI step doesnt have the same results.
> 

Hey @falcucci, I guess that you are probably running the previous version of the end to end test :thinking: 
Can you share the command that you are using to run the end to end test locally?

I have been able to reproduce the problem locally on my machine (it's the same problem as on the CI):
```
Error: Dependencies Builder can not create aggregator runner

Caused by:
    Dependency initialization error: «Error while building EraChecker» with additional nested error: 'Adapter Error message: «Reading from EraReader adapter raised an error: 'general error'.»
    
    Caused by:
        0: general error
        1: PallasChainObserver failed to get utxo 
        2: error while sending or receiving data through the channel
    
    Stack backtrace:
       0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
       1: mithril_aggregator::dependency_injection::builder::DependenciesBuilder::get_era_checker::{{closure}}
       2: mithril_aggregator::dependency_injection::builder::DependenciesBuilder::build_dependency_container::{{closure}}
       3: mithril_aggregator::commands::serve_command::ServeCommand::execute::{{closure}}
       4: mithril_aggregator::commands::MainCommand::execute::{{closure}}
       5: tokio::runtime::park::CachedParkThread::block_on
       6: mithril_aggregator::main
       7: std::sys_common::backtrace::__rust_begin_short_backtrace
       8: std::rt::lang_start::{{closure}}
       9: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:284:13
      10: std::panicking::try::do_call
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
      11: std::panicking::try
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
      12: std::panic::catch_unwind
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
      13: std::rt::lang_start_internal::{{closure}}
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:48
      14: std::panicking::try::do_call
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
      15: std::panicking::try
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
      16: std::panic::catch_unwind
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
      17: std::rt::lang_start_internal
                 at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:20
      18: std::rt::lang_start
      19: __libc_start_call_main
                 at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
      20: __libc_start_main_impl
                 at ./csu/../csu/libc-start.c:392:3
      21: _start'.
```

**Comment by falcucci** [comment]
@jpraynaud this error is the fix that I've mentioned that I already made and it's working here, but cant reproduce anymore.

is there any change that the aggregator isnt being updated with the bugfix in the CI?

**Comment by falcucci** [comment]
@jpraynaud just to make it documented here: got it [running](https://github.com/txpipe/mithril/actions/runs/7640787308?pr=1) changing the node version to 8.1.2 

the local state queries changed between node versions, so using 8.7.3 breaks it since pallas has not support for now.

**Comment by falcucci** [comment]
@jpraynaud FYI this has been addressed at https://github.com/txpipe/pallas/issues/385 and a proposed fix at https://github.com/txpipe/pallas/pull/386

**Comment by jpraynaud** [comment]
Hi @falcucci, now it's working with `8.7.3` but it seems that there is a backward compatibility problem with `8.1.2`. 

Here is the job that fails in the CI: https://github.com/input-output-hk/mithril/actions/runs/7668627456/job/20902207802

**Comment by falcucci** [comment]
@jpraynaud it wasn't upstream updated, fixed :)

**Comment by jpraynaud** [comment]
> @jpraynaud it wasn't upstream updated, fixed :)
> 
> when we have a new pallas version i'll bump it and notify you

Great news @falcucci! Once the new pallas version is available, we'll merge the PR :tada: 

Thanks for the good work!

**Comment by falcucci** [comment]
@jpraynaud just updated the upstream and bumped the new pallas version.

**Comment by jpraynaud** [comment]
Hi @falcucci, can you rebase once more on the main branch and bump the patch version of `mithril-common`?
I'll merge the PR thereafter :rocket:

**Comment by falcucci** [comment]
@jpraynaud done :)
